### ISSUE TITLE
[concepts] deferred substitution into requirements of class template members not implemented

### ISSUE THREAD

**Issue by marehr** [llvm/llvm-project#44178]
|  |  |
| --- | --- |
| Bugzilla Link | [44833](https://llvm.org/bz44833) |
| Version | trunk |
| OS | Linux |
| CC | @CaseyCarter,@erichkeane,@friedkeenan,@h-2,@JohelEGP,@AlexeyDmitriev,@PiliLatiesa,@zygoloid,@saarraz,@HighCommander4,@jwakely |

## Extended Description 
Sorry for this verbose and hard to read code, but I wanted to make this configurable.

This code basically has multiple code paths. If REQUIRES_BUG is set, the impl function uses requires to infer if char_is_valid_for is defined or not. If not impl uses SFINAE.

```c++
#include <utility>

// #define REQUIRES_BUG     1 - enable the requires bug, 0 - disable it
// #define FUNCTION_DEFINED 1 - define function char_is_valid_for, 0 - do not define char_is_valid_for 

template <int I> struct priority_tag : priority_tag<I - 1> {};
template <> struct priority_tag<0> {};

struct alphabet_base {};

#if FUNCTION_DEFINED
using expect_return_type = bool;
bool char_is_valid_for(char, alphabet_base) { return true; };
#else
using expect_return_type = int;
#endif

namespace seqan3::detail::adl_only {

template <typename...> void char_is_valid_for(...) = delete;

template <typename alph_t>
struct char_is_valid_for_fn
{
    using s_alph_t = alph_t;

#if REQUIRES_BUG
    template <typename t>
    static decltype(auto) impl(priority_tag<1>, t v)
        requires requires { (char_is_valid_for(v, s_alph_t{})); }
    { return (char_is_valid_for(v, s_alph_t{})); }
#else
    template <typename t>
    static auto impl(priority_tag<1>, t v)
        -> std::enable_if_t<true, decltype(char_is_valid_for(v, s_alph_t{}))>
    { return (char_is_valid_for(v, s_alph_t{})); }
#endif

    template <typename t, typename = void>
    static decltype(auto) impl(priority_tag<0>, t v) { return 0; }
};
} // namespace seqan3::detail::adl_only

int main() {
  auto a = seqan3::detail::adl_only::char_is_valid_for_fn<alphabet_base>::impl(priority_tag<1>{}, char{});
  static_assert(std::is_same_v<decltype(a), expect_return_type>);
}
```

As you can see on https://godbolt.org/z/tyeiJ2 when defining `clang++ -std=c++2a -DREQUIRES_BUG=1 -DFUNCTION_DEFINED=0` the requires block should silently see that the requires is not valid (because of a deleted function) and skip to the next definition.

Error Message:

```
<source>:31:15: error: call to deleted function 'char_is_valid_for'
    { return (char_is_valid_for(v, s_alph_t{})); }
              ^~~~~~~~~~~~~~~~~

<source>:45:75: note: in instantiation of function template specialization 'seqan3::detail::adl_only::char_is_valid_for_fn<alphabet_base>::impl<char>' requested here

  auto a = seqan3::detail::adl_only::char_is_valid_for_fn<alphabet_base>::impl(priority_tag<1>{}, char{});
                                                                          ^

<source>:20:29: note: candidate function [with $0 = <>] has been explicitly deleted
template <typename...> void char_is_valid_for(...) = delete;
                            ^

<source>:45:8: error: variable has incomplete type 'void'
  auto a = seqan3::detail::adl_only::char_is_valid_for_fn<alphabet_base>::impl(priority_tag<1>{}, char{});
       ^

2 errors generated.
```

Interestingly, if you substitute `s_alph_t` with `alph_t;` everything works as expected.

**Comment by erichkeane** [comment]
> Is there any progress on this issue? Does it indeed break all usage of views from libstdc++?

Sorry @h-2 that this took a while to respond to.  My understanding is yes, this breaks a significant portion of the 'ranges' views (and much of the rest of ranges as well).  I looked at it for a while and made only a minimum amount of progress, not enough to consider it 'progress'.  I can disable the premature instantiation of the concept, but cannot figure out how to properly re-instantiate it when it needs to be done.

I suspect someone who understands the template instantiation engine better than I (like @zygoloid ) would have to either give some extremely good hints, or just do it themselves :(

**Comment by h-2** [comment]
Thanks for the update! It seems like this is really a showstopper for lots of C++20 goodness, but there is not much I can do to help. If you do get any experimental code, I would of course be willing to test it.

**Comment by erichkeane** [comment]
As an update... I'm still working on this as my non-emergency-top-priority type thing.  I FEEL like I'm making progress, but I've still got 5 tests that crash.

**Comment by erichkeane** [comment]
FWIW, I was able to get by all of the lit tests today on this patch and tested a handful of ranges examples (including the one from @llvmbot here) https://reviews.llvm.org/D119544.

**Comment by erichkeane** [comment]
For everyone's information: Yesterday I concluded that it is as good as I think I can get it, so I opened it for review/added reviewers: https://reviews.llvm.org/D119544 

It passes the examples here, and all the lit tests as well as every ranges example I could come up with (at least as good as the old version).

**Comment by PiliLatiesa** [comment]
Hi @erichkeane 

Does your patch fix this testcase?

```c++
#include <concepts>

template<typename T, typename U>
struct foo
{
  void bar() const requires std::same_as<T, int> {} // #1

  void bar() const requires (std::same_as<T, int> && std::same_as<U, int>) {} // Subsumes #1
};

int main()
{
  foo<int, int> x;
  x.bar();
}
```

Thanks

**Comment by erichkeane** [comment]
> Hi @erichkeane
> 
> Does your patch fix this testcase?
> 
> ```c++
> #include <concepts>
> 
> template<typename T, typename U>
> struct foo
> {
>   void bar() const requires std::same_as<T, int> {} // #1
> 
>   void bar() const requires (std::same_as<T, int> && std::same_as<U, int>) {} // Subsumes #1
> };
> 
> int main()
> {
>   foo<int, int> x;
>   x.bar();
> }
> ```
> 
> Thanks, Pili

I wouldn't expect it to, I didn't touch any of the subsumption calculations.  HOWEVER, I ended up having to revert my patch a few times, and discovered my approach was likely misguided.  I am trying to come up with a different way of implementing the deferred substitution rule.

**Comment by PiliLatiesa** [comment]
Thanks @erichkeane 

I asked because `bar` being a member function appears to make a difference:

```c++
template<typename>
constexpr bool b = true;

template<typename T>
concept C = b<T>;

template<typename T, typename U>
class foo
{
public:
  void bar() const requires C<T> {}

  void bar() const requires (C<T> && C<U>) {}
};

template<typename T, typename U>
void bar(foo<T, U> const &) requires C<T> {}

template<typename T, typename U>
void bar(foo<T, U> const &) requires (C<T> && C<U>) {}

int main()
{
  foo<int, int> x;
  
  x.bar();

  bar(x);  // Clang does not complain

  return 0;
}
```
Thanks,
Pili

**Comment by h-2** [comment]
I see that some of your code has been merged, is that correct?

Should we report any problems we encounter here or open separate issues? Or is this "WIP" and you don't want further reports at this moment?

Thanks again for working on this!

**Comment by erichkeane** [comment]
It hasn't been merged, we tried a few times and it got reverted each time for breaking something else.  I still feel like I'm "close", but still working on it :/

**Comment by erichkeane** [comment]
Whew: https://github.com/llvm/llvm-project/commit/d483730d8c3fa2e0d4192b2f3c61c761b124e6ad

**Comment by erichkeane** [comment]
> Whew: [d483730](https://github.com/llvm/llvm-project/commit/d483730d8c3fa2e0d4192b2f3c61c761b124e6ad)

welp. I may have spoken too soon: https://reviews.llvm.org/rGd483730d8c3f

I don't have the reproducer yet, but if that ends up being any sizable amount of work, I may have to revert this... I even waited until it had made it 2 whole days :/

**Comment by erichkeane** [comment]
Apparently someone at facebook discovered this causes a regression in their internal build, so i was forced to revert.

**Comment by Predelnik** [comment]
Shouldn't this be reopen then?

**Comment by erichkeane** [comment]
Yep, could have sworn I did.  Thanks for the reminder.

**Comment by erichkeane** [comment]
Patch re-committed as yesterday, but I'm not going to be brave enough to mark this 'closed' for a few more days at minimum :)

**Comment by erichkeane** [comment]
Alright, we made it a week now and only have 2 different regressions reported.  One was fixed immediately, the other (#57958) is under review.  So I think I'm far enough along to have SOME faith this isn't going to be reverted.

Thanks all for the help with testing, and the reports in the first place!

**Comment by erichkeane** [comment]
> @erichkeane any chance of this getting back-ported to LLVM/Clang 13?

None whatsoever unfortunately.  There have been a number of regressions thanks to this (that I've been working through as they come up), and I'm sure there are likely more.  This would be too high risk to backport to any of the 13-15 branches.

So if you want to use this, you'll have to update to Clang 16 when it comes out.

**Comment by dimitry-ishenko** [comment]
Thanks @erichkeane. I didn't think so, but I figured it won't hurt to ask. 🤷🏻‍♂️ 😃

**Comment by kelbon** [comment]
> Alright, we made it a week now and only have 2 different regressions reported. One was fixed immediately, the other (#57958) is under review. So I think I'm far enough along to have SOME faith this isn't going to be reverted.
> 
> Thanks all for the help with testing, and the reports in the first place!

It still dont work for `= default` methods and destructors as i see

https://godbolt.org/z/r77oWoMTr

May be in can be easily fixed (based on previous fix)?

**Comment by erichkeane** [comment]
> > Alright, we made it a week now and only have 2 different regressions reported. One was fixed immediately, the other (#57958) is under review. So I think I'm far enough along to have SOME faith this isn't going to be reverted.
> > Thanks all for the help with testing, and the reports in the first place!
> 
> It still dont work for `= default` methods and destructors as i see
> 
> https://godbolt.org/z/r77oWoMTr
> 
> May be in can be easily fixed (based on previous fix)?

Interesting, thanks for  pointing this out!  @royjacobson was poking around default function constraints at one point as well, so he might have some idea.

**Comment by erichkeane** [comment]
I see that the code that calls this is part of `SetEligibleMethods` which @royjacobson wrote a while back, and is intending to implement class.mem.special p6.  I don't have a good idea what all is going on there, nor if it was supposed to have been 'eliminated' by that in the first place.  I'm hoping Roy can comment.

**Comment by royjacobson** [comment]
The standard is clear that we can't defer constraints checks for destructors, but P0848 is unclear about when, or if, to perform the constraints checks for the other special member functions.

When I implemented P0848 I decided to preemptively evaluating the constraints when we complete a class, and that it's probably OK. Not doing it would've added a non-trivial effort to support deferred computation of `canPassInRegs` and class [copy-]triviality, which we currently store as flags in `RecordDeclBits`. AFAIK pretty much everywhere assumes that those properties are fixed once a class is complete, and changing that seemed to me something worthy to avoid, as long as the standard doesn't explicitly prevents us from doing it.

But maybe this is non compliant and we really should defer the triviality calculations.

**Comment by erichkeane** [comment]
Hmm... that seems worthy of a CWG reflector discussion?  Mind starting one?

**Comment by zygoloid** [comment]
I don't think we should, or reasonably can, defer these calculations in general. This seems to be the same kind of situation as in CWG1344, which described another way in which the set of special member functions could change after the class becomes complete.
